engineioxide/src/engine.rs

```diff
@@ -153,39 +186,54 @@ impl<H: EngineIoHandler> EngineIo<H>
             debug!("error aggregating body: {:?}", e);
             Error::HttpErrorResponse(StatusCode::BAD_REQUEST)
         })?;
-        let packets = body.reader().split(b'\x1e');
-
+        
         let socket = self
             .get_socket(sid)
             .ok_or(Error::UnknownSessionID(sid))
             .and_then(|s| s.is_http().then(|| s).ok_or(Error::TransportMismatch))?;
 
-        for packet in packets {
-            match Packet::try_from(packet?) {
+        let packets = Payload::new(protocol, body.reader());
+
+        for p in packets {
+            let raw_packet = p.map_err(|e| {
+                debug!("error parsing packets: {:?}", e);
+                self.close_session(sid);
+                Error::HttpErrorResponse(
+                    StatusCode::BAD_REQUEST,
+                )
+            })?;
+
+            let packet = match Packet::try_from(raw_packet) {
+                Ok(p) => p,
                 Err(e) => {
                     debug!("[sid={sid}] error parsing packet: {:?}", e);
                     self.close_session(sid);
-                    Err(e)
+                    return Err(e);
                 }
-                Ok(Packet::Close) => {
+            };
+
+            match packet {
+                Packet::Close => {
                     debug!("[sid={sid}] closing session");
                     socket.send(Packet::Noop)?;
                     self.close_session(sid);
                     break;
                 }
-                Ok(Packet::Pong) => socket
-                    .pong_tx
-                    .try_send(())
-                    .map_err(|_| Error::HeartbeatTimeout),
-                Ok(Packet::Message(msg)) => {
+                Packet::Pong | Packet::Ping => {
+                    socket
+                        .pong_tx
+                        .try_send(())
+                        .map_err(|_| Error::HeartbeatTimeout)
+                },
+                Packet::Message(msg) => {
                     self.handler.on_message(msg, &socket);
                     Ok(())
                 }
-                Ok(Packet::Binary(bin)) => {
+                Packet::Binary(bin) | Packet::BinaryV3(bin) => {
                     self.handler.on_binary(bin, &socket);
                     Ok(())
                 }
-                Ok(p) => {
+                p => {
                     debug!("[sid={sid}] bad packet received: {:?}", &p);
                     Err(Error::BadPacket(p))
```

Comment:
is it required to split into Packet::try_from and match?

Reply:
Good point. I think I split it while implementing the iterator but on second thought it's probably not necessary. I'll see if I can combine them again.